src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -97,7 +97,14 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.semaphore.Dispose();
+                // !!DO NOT!! dispose the semaphore, for the following reasons:
+                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
+                //    requires tracking of all outstanding requests for enter/release.
+                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
+                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
+                //    do not access this property.
```

Comment:
> Ideally it's easy to explain.

I mostly agree with your "ideal", although I'm not sure if throwing `OperationCanceledException` would be more appropriate than `ObjectDisposedException`. If you call `EnterAsync` after a disposal or concurrently with an in-progress disposal, an `ObjectDisposedException` would arguably be the expected result. So introducing an edge race condition where we might throw `OperationCanceledException` instead (when the disposal was complete before the `EnterAsync` call hit the check) sounds like introducing more room for errors. And folks may not expect that exception type at all if they didn't supply a `CancellationToken` when they called `EnterAsync`. 
And anyway, if we throw *any* exception here, Roslyn might still crash unless they're expecting the exception. And if they're OK with an exception, can't they catch `ObjectDisposedException`?

> For the same reason AsyncSemaphore exists in the first place

I don't follow that an async wrapper around `SemaphoreSlim` should allow callers to dispose the very thing it's also using in another context when `SemaphoreSlim` doesn't allow it. The reason `AsyncSemaphore` was created was to make a more convenient syntax for asynchronously awaiting and ensuring a Dispose call is made. It wasn't so folks could Dispose something they're still using.

I think if Roslyn would be willing to no-op on Dispose, they should just not call Dispose on their side.